File: fixtures/java/j09/tests/src/test/java/OrderServiceTest.java
// Tests for OrderService — verifies that the null-pointer bug is fixed
// Functionally equivalent to the Node.js order-service.test.js

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class OrderServiceTest {

    @Test
    void successfullyCreatesOrderForExistingProduct() {
        Order order = new Order("prod-1", 2);
        OrderResult result = OrderService.processOrder(order);

        assertNotNull(result, "Result should not be null");
        assertNotNull(result.getOrderId(), "Order ID should be set");
        assertTrue(result.getOrderId().startsWith("ORD-"), "Order ID should start with ORD-");
        assertEquals(59.98, result.getTotal(), 0.01, "Total should be 29.99 * 2 = 59.98");
        assertEquals("confirmed", result.getStatus(), "Status should be confirmed");
    }

    @Test
    void throwsDescriptiveErrorForMissingProduct() {
        Order order = new Order("non-existent-product", 1);

        Exception exception = assertThrows(Exception.class, () -> {
            OrderService.processOrder(order);
        });

        String message = exception.getMessage().toLowerCase();
        assertTrue(
            message.contains("not found") || message.contains("does not exist") || message.contains("product"),
            "Error message should mention product not found, got: " + exception.getMessage()
        );
    }

    @Test
    void doesNotThrowNullPointerForMissingProduct() {
        Order order = new Order("does-not-exist", 1);

        Exception exception = assertThrows(Exception.class, () -> {
            OrderService.processOrder(order);
        });

        // The fix should throw a descriptive error, NOT a NullPointerException
        assertFalse(exception instanceof NullPointerException,
                "Should not throw NullPointerException — should have a proper null check");
    }

    @Test
    void processOrderReturnsCorrectTotal() {
        Order order = new Order("prod-2", 3);
        OrderResult result = OrderService.processOrder(order);

        assertEquals(149.97, result.getTotal(), 0.01, "Total should be 49.99 * 3 = 149.97");
        assertEquals("confirmed", result.getStatus(), "Status should be confirmed");
    }
}

File: fixtures/nodejs/j09/tests/order-service.test.js
const request = require('supertest');
const { app, processOrder } = require('./order-service');

describe('Order Service - Null check regression', () => {
  test('successfully creates order for existing product', async () => {
    const res = await request(app).post('/orders').send({
      productId: 'prod-1',
      quantity: 2,
    });
    expect(res.status).toBe(201);
    expect(res.body).toHaveProperty('orderId');
    expect(res.body.total).toBe(59.98);
  });

  test('returns 404 or 400 for non-existent product (not 500)', async () => {
    const res = await request(app).post('/orders').send({
      productId: 'non-existent-product',
      quantity: 1,
    });
    expect(res.status).not.toBe(500);
    expect([400, 404]).toContain(res.status);
  });

  test('processOrder throws descriptive error for missing product', async () => {
    await expect(processOrder({ productId: 'does-not-exist', quantity: 1 }))
      .rejects.toThrow(/not found|does not exist|product/i);
  });

  test('processOrder returns correct total', async () => {
    const result = await processOrder({ productId: 'prod-2', quantity: 3 });
    expect(result.total).toBe(149.97);
    expect(result.status).toBe('confirmed');
  });
});
